libslide/src/scanner.rs

```diff
@@ -0,0 +1,209 @@
+        let t = match c {
+            '+' => TokenType::Plus,
+            '-' => TokenType::Minus,
+            '*' => TokenType::Mult,
+            '/' => TokenType::Div,
+            '%' => TokenType::Mod,
+            '^' => TokenType::Exp,
+            '=' => TokenType::Equal,
+            '(' => TokenType::OpenParen, 
+            ')' => TokenType::CloseParen, 
+            '[' => TokenType::OpenBracket,
+            ']' => TokenType::CloseBracket,
+             _  => TokenType::Invalid(c.to_string())
+        };
+        let ret = Token{token: t};
+        return ret;
+   }
+
+    // iterates through any digits to create a token of that value
+    fn iterate_digit(&mut self,mut i: usize) -> (Token, usize){
+        let mut int_str = "".to_owned();
+        let mut dec_str = ".".to_owned();
+        let ret: Token;
+        // iterate through integer part
+        while i < self.input.chars().count() && (self.input.as_bytes()[i] as char).is_digit(10){
+            int_str.push(self.input.as_bytes()[i] as char);
+            i += 1;
+        }
+        // iterate through decimal
+        if i < self.input.chars().count() && (self.input.as_bytes()[i] as char)== '.'{
+            i += 1;
+            while i < self.input.chars().count() && (self.input.as_bytes()[i] as char).is_digit(10){
+                dec_str.push(self.input.as_bytes()[i] as char);
+                i += 1;
+            }
+            int_str.push_str(&dec_str);
+            // turn integer and decmial strings into token
+            ret = Token{token: TokenType::Float(int_str.parse::<f64>().unwrap())}
+        }
+        else{
+            // turn integer string into token and default the float
+            ret = Token{token:TokenType::Int(int_str.parse::<i64>().unwrap())}
+        }
+        return (ret, i);
+    }
+
+    pub fn scan(&mut self){
+        let mut i: usize = 0;
+        let mut c: char;
+        let mut t: Token;
+        let mut tuple: (Token, usize);
+        // iterate through string
+        while i < self.input.chars().count() {
+            // ignore whitespace
+            if !((self.input.as_bytes()[i] as char).is_whitespace()) {
+                // check for digit and call correct helper function 
+                if (self.input.as_bytes()[i] as char).is_digit(10) {
+                    tuple = Scanner::iterate_digit(self, i);
+                    i = tuple.1;
+                    self.output.push(tuple.0);
+                }
+                else{
+                    self.output.push(Scanner::create_symbol_token(self.input.as_bytes()[i] as char));
+                    i += 1;
+                }       
+            }
+            else{
+                i += 1;
+            }
+        }
+    }
+    
+}
+
+#[cfg(test)]
+mod tests{
+    use super::*;
+
+    fn compare_vec<T: PartialEq>(a: &Vec<T>, b: &Vec<T>) -> bool{
+        let equal = a.iter().zip(b.iter()).filter(|&(a,b)| a == b).count();
+        equal == a.len() && equal == b.len()
+    }
```

Comment:
I think rust already supports direct comparison of two vectors..

Reply:
yes will fix.